### PR TITLE
Kernel region and omp thread barrier statements

### PR DESCRIPTION
**Pull request by artv3**
This PR enables users to create OpenMP parallel regions inside kernel and adds OpenMP sync threads as a statement.

### CODE REVIEW

File: test/unit/test-kernel.cpp [hunk 1]
```diff
@@ -918,7 +918,129 @@ TEST(Kernel, CollapseSeq)
...
+
+  using Pol =
+    RAJA::KernelPolicy<
+      RAJA::statement::Region<RAJA::seq_region,
+
+        RAJA::statement::For<0, RAJA::loop_exec,
+                           RAJA::statement::Lambda<0>
+        >
+
+       ,RAJA::statement::For<0, RAJA::loop_exec,
```
Comment (rhornung67) [inline test/unit/test-kernel.cpp]: Please format this policy so it's easier to read; e.g., indentation, vertical alignment, and move commas to end of preceding line.

File: test/unit/test-kernel.cpp [hunk 2]
```diff
@@ -918,7 +918,129 @@ TEST(Kernel, CollapseSeq)
...
+    Arr_a[i] = 0;
+    Arr_b[i] = 0;
+    Arr_c[i] = 0;
+  }
+
+  using Pol =
+    RAJA::KernelPolicy<
+      RAJA::statement::Region<RAJA::omp_parallel_region,
+
+        RAJA::statement::For<0, RAJA::omp_for_nowait_exec,
```
Comment (rhornung67) [inline test/unit/test-kernel.cpp]: Same comment about formatting policy.

**Comment by artv3** [comment]
Something is fishy here, the unit test seems to fail with GNU 4.9 and 7.3.1 in debug mode. But it works correctly with the release version. @trws , I was wondering if you might have any suggestions?

**Comment by trws** [comment]
Oh boy... I have a bad feeling about this.  Now it’s the clang debug build, something is very fishy here.

**Comment by rhornung67** [comment]
@artv3 This suggestion is really grasping at straws, but have you tried using a tuple of three range segments (all the same size) and using a different one for each 'For' statement? This is not the same scenario, but I have run into issues in the perf suite when I tried to use the same range segment in multiple loops in a kernel. Anyway, it may be worth a try. If that works, then it may help reveal the issue...

**Comment by artv3** [comment]
@rhornung67 , I gave the idea a spin but still ran into the same issues. Correct results for the optimized version, but incorrect for the optimized version.

**Comment by rhornung67** [comment]
@artv3 I thought it was a long shot. Thanks for trying. Have you tried stepping through the code with a debugger?

**Comment by artv3** [comment]
@MrBurmark and I made a couple of changes to the internals of Region. It looks like we needed to make thread private copies of the loop data. This PR is in good shape now.

**Pull request merged**